packages/shared-models/src/ymodels.ts

```diff
@@ -26,6 +26,44 @@ export interface IYText extends models.ISharedText {
 export type YCellType = YRawCell | YCodeCell | YMarkdownCell;
 
 export class YDocument<T> implements models.ISharedDocument {
+  /**
+   * Get/Set the isDisposed attribute to know whether this model
+   * has been disposed or not.
+   */
+  get isDisposed(): boolean {
+    return this._isDisposed;
+  }
+  set isDisposed(value: boolean) {
+    this._isDisposed = value;
+  }
+
+  /**
+   * The Yjs document.
+   */
+  readonly ydoc = new Y.Doc();
+
+  /**
+   * The Yjs document.
+   */
+  readonly ysource = this.ydoc.getText('source');
+
+  /**
+   * The awareness attribute to access the user's state.
+   */
+  readonly awareness = new Awareness(this.ydoc);
+
+  /**
+   * The undoManager that handles the change history.
+   */
+  readonly undoManager = new Y.UndoManager([this.ysource], {
+    trackedOrigins: new Set([this])
+  });
+
+  /**
+   * The changed signal.
+   */
+  readonly changed = new Signal<this, T>(this);
```

Comment:
Do we need to expose a `Signal` instead of an `ISignal`?

Reply:
No, we should expose an ISignal.
Would this work?
```ts
readonly changed: ISignal<this, T> = new Signal<this, T>(this);
```
The problem is that we need to create the signal before the object is initialized.

Reply:
> The problem is that we need to create the signal before the object is initialized.

Does it mean this field should be initialized before the constructor runs?

For reference, and to get a clearer idea of why this is needed, which component(s) depends on that behavior? (probably they have a `.sharedModel.changed.connect()` somewhere)

Reply:
`YDocument<T>` is the base class of `YFile` and `YNotebook` so we need to create the `Signal` before the constructor because of `this` and `T`.

The only class that uses this signal is the `Model` on `codeEditor`: https://github.com/jupyterlab/jupyterlab/blob/dd992bae99f49a279f6b68b864711d04a796d682/packages/codeeditor/src/editor.ts#L240